File: go/storage.go
// Package storage is where bytes live, addressed by what they are rather than
// by what somebody decided to call them.
//
// # The problem, measured rather than assumed
//
// One ordinary machine, four tools, four stores that know nothing about each
// other:
//
//	LM Studio          44 GB   names files by filename only   NOT dedupable
//	Lemonade           26 GB   content hash
//	Ollama             23 GB   content hash
//	HuggingFace cache  23 GB   content hash
//
// 116 GB, and the same weights sit in more than one of them under different
// names. Not because the problem is hard — because there is no shared interface,
// so each tool solves it privately and none of them solve it well. This project
// opened by measuring that and then spent months not building the layer.
//
// # What an interface here has to be able to say
//
// Two things, and the second is the one that was missing:
//
//	Find   are these exact bytes already on this machine?
//	Place  where should bytes we do not have yet go?
//
// Find is what makes dedup possible: once a registry has told us an artifact's
// digest, discovering that Ollama already has it is a stat, not a download. That
// half already existed, in the model layer, reading three real stores.
//
// Place is what stops an application inventing a path. Before it,
// download.Sink{Partial, Final} were two strings every caller filled in by
// hand — so the layer that is pluggable about WHO fetches the bytes was
// hardcoded about WHERE they land, which is the same defect at the other end of
// one call.
//
// # What this must not become
//
// Not a filesystem. A Ref is opaque, and a store that happens to be a directory
// says so through the Local capability rather than by returning paths from the
// interface. That is the same shape job.Scratch has, for the same reason: the
// binding must not be able to name itself upward.
package storage

import "errors"

// Ref is a handle to bytes in a store. Opaque on purpose.
//
// The application that asked for a download does not learn a path from it, and
// could not act on one if it did — the process that finally writes those bytes
// may be a Windows service under its own account, or a daemon on a NAS that
// mounts this store somewhere else entirely.
type Ref struct {
	// Store names which store this came from, so a caller holding several refs
	// can say where each one is without asking.
	Store string
	// Digest is what the bytes are, "sha256:<hex>", when it is known. A store
	// may hold bytes whose digest nobody has computed.
	Digest string
	// Size is how big, or 0 when unknown.
	Size int64

	// locator is how the store's own binding finds it again. Unexported: a
	// caller that could read it would start joining paths onto it.
	locator string
}

// Store is somewhere bytes live.
//
// Deliberately small. Everything a store must be able to do is here; everything
// only SOME stores can do is a capability below.
type Store interface {
	// Name is how this store is identified to a person. "ollama", "nas".
	Name() string

	// Find reports where these bytes already are. The digest is authoritative;
	// a store that cannot answer by digest cannot participate, and LM Studio —
	// the largest store on the machine that was measured — is exactly that case.
	//
	// It must not hash anything. A hit is trusted only as far as the store's own
	// naming convention goes, which is why whoever copies from it still
	// verifies: Ollama's blobs hash to their own filenames, and Ollama does not
	// check them, so a corrupt blob there is a real possibility.
	Find(digest string) (Ref, bool)

	// Place reserves somewhere for bytes this store does not have yet.
	//
	// Reserving is not writing. The caller may be handing this to a service that
	// will do the writing much later, under a different account, after this
	// process has exited — so Place settles WHERE and nothing else.
	Place(digest string, size int64) (Ref, error)
}

// Local is an OPTIONAL capability: a store whose binding is a filesystem can
// name a location on it.
//
// It exists because bytes are not always moved by us. BITS writes under its own
// service account and hands the file over on completion; a NAS daemon writes on
// the far side of a share. Neither can be given an io.Writer, so a destination
// they can act on has to be expressible — and that destination is a path.
//
// It is a capability rather than part of Store for the same reason job.Scratch
// is: a store backed by a service answers no, and a caller then has to have a
// real answer for that case instead of assuming a directory exists.
type Local interface {
	// Path is where this ref is on this machine's filesystem.
	Path(Ref) string
}

// Writable is an OPTIONAL capability: a store that can take bytes directly,
// for a caller that is holding them rather than delegating the transfer.
type Writable interface {
	// Commit moves already-written bytes into the store under their digest,
	// making them findable. Until this is called, a placed ref is a reservation
	// and nothing more — which is the same two-phase shape the job layer uses
	// for TRANSFERRED, and for the same reason: the writer and the consumer are
	// not the same process.
	Commit(Ref) error
}

var (
	// ErrReadOnly is returned by Place on a store that can only be read. The
	// foreign stores are all like this: we look inside Ollama's blobs, and we
	// do not write there, because their layout is theirs to change.
	ErrReadOnly = errors.New("storage: this store is read-only")
	// ErrNotFound means these bytes are not here.
	ErrNotFound = errors.New("storage: not found")
)

// Locator returns a ref's internal location. For the store that produced it.
//
// Exported for bindings, not for applications: it is how a Local implementation
// turns a ref back into something it can act on. An application that reaches for
// this is doing the thing this package exists to stop.
func Locator(r Ref) string { return r.locator }

// NewRef is how a Store constructs a ref. Also for bindings.
func NewRef(store, digest, locator string, size int64) Ref {
	return Ref{Store: store, Digest: digest, Size: size, locator: locator}
}

File: cpp/src/storage.cpp
#include <abstraction/storage/storage.h>

#include <algorithm>
#include <cctype>
#include <filesystem>
#include <system_error>

namespace fs = std::filesystem;

namespace abstraction {
namespace storage {

Ref make_ref(const std::string& store, const std::string& digest, const std::string& locator,
             std::int64_t size) {
    Ref r;
    r.store_ = store;
    r.digest_ = digest;
    r.locator_ = locator;
    r.size_ = size;
    return r;
}

const std::string& locator_of(const Ref& r) { return r.locator_; }

std::string normal_digest(const std::string& raw) {
    std::string s;
    for (char c : raw) {
        if (c != ' ' && c != '\t' && c != '\n' && c != '\r') {
            s.push_back(static_cast<char>(std::tolower(static_cast<unsigned char>(c))));
        }
    }
    for (const char* prefix : {"sha256:", "sha256-"}) {
        const std::string p(prefix);
        if (s.rfind(p, 0) == 0) {
            s = s.substr(p.size());
            break;
        }
    }
    if (s.size() != 64) {
        return "";
    }
    for (char c : s) {
        const bool hex = (c >= '0' && c <= '9') || (c >= 'a' && c <= 'f');
        if (!hex) {
            return "";
        }
    }
    return "sha256:" + s;
}

namespace {

// "sha256:<hex>" as a filename. Hyphen and not colon, because a colon is not a
// legal character in an NTFS filename — and because this is the spelling Ollama
// already uses, so its blobs directory and this one are the same shape.
std::string blob_name(const std::string& normalized) {
    std::string out = normalized;
    const std::size_t colon = out.find(':');
    if (colon != std::string::npos) {
        out[colon] = '-';
    }
    return out;
}

}  // namespace

ContentStore::ContentStore(const std::string& name, const std::string& root) : name_(name), root_(root) {
    if (root.empty()) {
        throw StorageError("storage: a content store needs a root");
    }
    std::error_code ec;
    fs::create_directories(fs::u8path(root) / "blobs", ec);
    fs::create_directories(fs::u8path(root) / "incoming", ec);
    if (ec) {
        throw StorageError("storage: could not prepare " + root + ": " + ec.message());
    }
}

bool ContentStore::find(const std::string& digest, Ref& out) const {
    const std::string normalized = normal_digest(digest);
    if (normalized.empty()) {
        return false;
    }
    const fs::path blob = fs::u8path(root_) / "blobs" / blob_name(normalized);
    std::error_code ec;
    const auto size = fs::file_size(blob, ec);
    if (ec) {
        return false;
    }
    out = make_ref(name_, normalized, blob.u8string(), static_cast<std::int64_t>(size));
    return true;
}

Ref ContentStore::place(const std::string& digest, std::int64_t size) {
    const std::string normalized = normal_digest(digest);
    if (normalized.empty()) {
        // A content-addressed store cannot name a location for bytes nobody can
        // identify. Refusing is the point: inventing a name here is exactly the
        // behaviour that produced 116 GB of duplicates across four tools.
        throw StorageError("storage: place needs a digest, got " + digest);
    }

    // Already here: return the bytes rather than a reservation, so asking twice
    // does not start a second transfer. Same rule the download service applies
    // to jobs, one layer down.
    Ref existing;
    if (find(normalized, existing)) {
        return existing;
    }
    const fs::path incoming = fs::u8path(root_) / "incoming" / blob_name(normalized);
    return make_ref(name_, normalized, incoming.u8string(), size);
}

std::string ContentStore::path(const Ref& r) const {
    const std::string normalized = normal_digest(r.digest());
    if (normalized.empty()) {
        throw StorageError("storage: this ref carries no digest");
    }
    return (fs::u8path(root_) / "blobs" / blob_name(normalized)).u8string();
}

std::string ContentStore::incoming_path(const Ref& r) const {
    const std::string normalized = normal_digest(r.digest());
    if (normalized.empty()) {
        throw StorageError("storage: this ref carries no digest");
    }
    return (fs::u8path(root_) / "incoming" / blob_name(normalized)).u8string();
}

void ContentStore::commit(const Ref& r) {
    const std::string normalized = normal_digest(r.digest());
    if (normalized.empty()) {
        throw StorageError("storage: this ref carries no digest");
    }
    const fs::path incoming = fs::u8path(incoming_path(r));
    const fs::path blob = fs::u8path(path(r));

    std::error_code ec;
    if (!fs::exists(incoming, ec)) {
        // Already committed is success: a second commit of the same bytes is a
        // retry, not an error, and the caller cannot always know which it is.
        if (fs::exists(blob, ec)) {
            return;
        }
        throw NotFound("storage: nothing placed at " + incoming.u8string() + " to commit");
    }

    // Rename, so the blob appears under its final name only once it is whole.
    // Anything less and a half-written blob sits there named by a digest it does
    // not have, and every other tool on this machine trusts the name.
    fs::rename(incoming, blob, ec);
    if (ec) {
        throw StorageError("storage: committing " + blob.u8string() + ": " + ec.message());
    }
}

}  // namespace storage
}  // namespace abstraction
